applications/wg-easy/taskfiles/container.yml

```diff
@@ -1,90 +1,111 @@
+      - task: start
+    cmds:
+      - echo "Connecting to {{.CONTAINER_NAME}}..."
+      - "{{.CONTAINER_RUNTIME}} exec -it {{.CONTAINER_NAME}} {{.CMD}}"
+    vars:
+      CMD: "{{if .CMD}}bash -c '{{.CMD}}'{{else}}/bin/bash{{end}}"
+
+  build:
+    run: once
     desc: Build development container image
     vars:
-      IMAGE_NAME: '{{.DEV_CONTAINER_REGISTRY}}/{{.DEV_CONTAINER_IMAGE}}:{{.DEV_CONTAINER_TAG}}'
+      IMAGE_NAME: "{{.CONTAINER_REGISTRY}}/{{.CONTAINER_IMAGE}}:{{.CONTAINER_TAG}}"
       CONTAINERFILE: '{{.CONTAINERFILE | default "./container/Containerfile"}}'
       BUILD_ARGS: '{{.BUILD_ARGS | default ""}}'
-    requires:
-      vars: [DEV_CONTAINER_REGISTRY, DEV_CONTAINER_IMAGE, CONTAINERFILE]
+      MESSAGE: "{{.MESSAGE}}"
+    cmds:
+      - '{{if .MESSAGE}}echo "{{.MESSAGE}}"{{end}}'
+      - "{{.CONTAINER_RUNTIME}} build -t {{.IMAGE_NAME}} -f {{.CONTAINERFILE}} ."
+    status:
+      - "{{.CONTAINER_RUNTIME}} image exists {{.IMAGE_NAME}}"
 
+  pull:
+    run: once
+    vars:
+      IMAGE_NAME: "{{.CONTAINER_REGISTRY}}/{{.CONTAINER_IMAGE}}:{{.CONTAINER_TAG}}"
+      CONTAINERFILE: '{{.CONTAINERFILE | default "./container/Containerfile"}}'
     cmds:
-      - '{{.CONTAINER_RUNTIME}} build -t {{.IMAGE_NAME}} -f {{.CONTAINERFILE}} .'
+      - cmd: "{{.CONTAINER_RUNTIME}} pull {{.IMAGE_NAME}}"
+        ignore_error: true
+    status:
+      - "{{.CONTAINER_RUNTIME}} image exists {{.IMAGE_NAME}}"
+
+  # Replaces long if/else commands with idempotent tasks and task-color user feedback
+  check-image-exists:
+    desc: If image does not exist, tries to pull. If pull fails, tries to build
+    run: always
+    vars:
+      IMAGE_NAME: "{{.CONTAINER_REGISTRY}}/{{.CONTAINER_IMAGE}}:{{.CONTAINER_TAG}}"
+    cmds:
+      - echo "Image '{{.IMAGE_NAME}}' does not exist. Trying pull"
+      - task: pull
+      - task: build
+        vars:
+          MESSAGE: "Pull failed. Trying build"
+    status:
+      - "{{.CONTAINER_RUNTIME}} image exists {{.IMAGE_NAME}}"
 
-  # Start development container in background.
   start:
     desc: Start development container in background
-    silent: true
     run: once
     vars:
-      IMAGE_NAME: '{{.DEV_CONTAINER_REGISTRY}}/{{.DEV_CONTAINER_IMAGE}}:{{.DEV_CONTAINER_TAG}}'
+      IMAGE_NAME: "{{.CONTAINER_REGISTRY}}/{{.CONTAINER_IMAGE}}:{{.CONTAINER_TAG}}"
       CONTAINERFILE: '{{.CONTAINERFILE | default "./container/Containerfile"}}'
       BUILD_ARGS: '{{.BUILD_ARGS | default ""}}'
     requires:
-      vars: [DEV_CONTAINER_REGISTRY, DEV_CONTAINER_IMAGE, DEV_CONTAINER_TAG, DEV_CONTAINER_NAME, REPLICATED_API_TOKEN]
-
-    status:
-      - '{{.CONTAINER_RUNTIME}} ps | grep -q "{{.DEV_CONTAINER_NAME}}"'
+      vars:
+        [
+          CONTAINER_REGISTRY,
+          CONTAINER_IMAGE,
+          CONTAINER_TAG,
+          CONTAINER_NAME,
+          REPLICATED_API_TOKEN,
+        ]
     cmds:
+      - task: check-image-exists
+      # Start with host networking for kubectl port-forward compatibility
```

Comment:
The security implications of this are a bit concerning, although if it's necessary it's not something I would loose sleep over.

Can you explain why kubectl port-forward is needed? Is this running port-forward from within the container, or external to the container? Is there a reason you can't just expose a port rather than use host networking?